akka-actor/src/main/scala/akka/actor/ActorSystem.scala

```diff
@@ -930,8 +943,23 @@ private[akka] class ActorSystemImpl(
   def registerOnTermination(code: Runnable): Unit = { terminationCallbacks.add(code) }
 
   override def terminate(): Future[Terminated] = {
-    if (!settings.LogDeadLettersDuringShutdown) logDeadLetterListener.foreach(stop)
-    guardian.stop()
+    val coordinatedShutdown = CoordinatedShutdown(this)
+    if (settings.CoordinatedShutdownRunByActorSystemTerminate &&
+        coordinatedShutdown.shutdownReason().isEmpty && !aborting) {
```

Comment:
If I understand correctly, this logic makes sure that the 'first' time `terminate()` is called, coordinated shutdown is triggered, while the 'second' time `terminate()` is called (presumably from the final CoordinatedShutdown phase) it 'actually' terminates the actor system.

That seems a bit brittle: what if `terminate()` is called twice from user code?

It might be easier to follow to introduce an internal `private[akka] def finalTerminate()` and call that one from the 'else' branch and from the final coordinated shutdown, making this `def terminate()` "idempotent", so to speak?

Reply:
That is correct.

Sure, I can add another `finalTerminate()` method, but that will not protect against multiple calls. My thinking was that if terminate() was called again while CoordinatedShutdown was still running the second call would be an eager shutdown.

Maybe we should add an AtomicBoolean here as well to make `terminate()` idempotent (only run once).

Reply:
My suggestion would be to remove the `coordinatedShutdown.shutdownReason().isEmpty` from this conditional - with that change I'd say it would protect against multiple calls, since `CoordinatedShutdown.run` is already protected?

"terminate() does a coordinated shutdown, unless it is called while coordinated shutdown is already running, then it does an eager shutdown" seems like a complicated API to me.

Reply:
I added the separate `finalTerminate`, see f0f0198
have some more thoughts, will comment on the latest diff